x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/ElasticsearchInternalService.java

```diff
@@ -1004,4 +993,80 @@ static TaskType inferenceConfigToTaskType(InferenceConfig config) {
             return null;
         }
     }
+
+    /**
+     * Iterates over the batch executing a limited number requests at a time to avoid
+     * filling the ML node inference queue.
+     *
+     * First, a single request is executed, which can also trigger deploying a model
+     * if necessary. When this request is successfully executed, a callback executes
+     * N requests in parallel next. Each of these requests also has a callback that
+     * executes one more request, so that at all time N requests are in-flight. This
+     * continues until all requests are executed.
+     */
+    class BatchIterator {
```

Comment:
logs from the BatchIterator:
```
[2024-10-25T17:04:47,457][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=0
[2024-10-25T17:04:47,797][INFO ][o.e.x.m.p.a.TransportLoadTrainedModelPackage] [runTask-0] [.elser_model_2] starting model import
[2024-10-25T17:05:06,425][INFO ][o.e.x.m.p.a.TransportLoadTrainedModelPackage] [runTask-0] [.elser_model_2] finished model import after [18] seconds
[2024-10-25T17:05:07,223][INFO ][stdout                   ] [runTask-0] ojAlgo includes a small set of predefined hardware profiles,
[2024-10-25T17:05:07,223][INFO ][stdout                   ] [runTask-0] none of which were deemed suitable for the hardware you're currently using.
[2024-10-25T17:05:07,223][INFO ][stdout                   ] [runTask-0] A default hardware profile, that is perfectly usable, has been set for you.
[2024-10-25T17:05:07,223][INFO ][stdout                   ] [runTask-0] You may want to set org.ojalgo.OjAlgoUtils.ENVIRONMENT to something that
[2024-10-25T17:05:07,223][INFO ][stdout                   ] [runTask-0] better matches the hardware/OS/JVM you're running on, than the default.
[2024-10-25T17:05:07,223][INFO ][stdout                   ] [runTask-0] Additionally it would be appreciated if you contribute your hardware profile:
[2024-10-25T17:05:07,224][INFO ][stdout                   ] [runTask-0] https://github.com/optimatika/ojAlgo/issues
[2024-10-25T17:05:07,224][INFO ][stdout                   ] [runTask-0] Architecture=aarch64 Threads=12 Memory=536870912
[2024-10-25T17:05:07,468][INFO ][o.e.x.m.i.d.DeploymentManager] [runTask-0] [.elser-2-elasticsearch] Starting model deployment of model [.elser_model_2]
[2024-10-25T17:05:13,149][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=1
[2024-10-25T17:05:13,149][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=3
[2024-10-25T17:05:13,149][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=2
[2024-10-25T17:05:13,149][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=4
[2024-10-25T17:05:13,149][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=5
[2024-10-25T17:05:13,150][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=7
[2024-10-25T17:05:13,149][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=6
[2024-10-25T17:05:13,150][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=8
[2024-10-25T17:05:13,150][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=9
[2024-10-25T17:05:13,151][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=10
[2024-10-25T17:05:13,151][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=12
[2024-10-25T17:05:13,151][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=13
[2024-10-25T17:05:13,151][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=14
[2024-10-25T17:05:13,151][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=15
[2024-10-25T17:05:13,152][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=16
[2024-10-25T17:05:13,152][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=17
[2024-10-25T17:05:13,151][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=11
[2024-10-25T17:05:13,152][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=19
[2024-10-25T17:05:13,152][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=18
[2024-10-25T17:05:13,152][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=20
[2024-10-25T17:05:15,403][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=21
[2024-10-25T17:05:17,292][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=22
[2024-10-25T17:05:18,439][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=23
[2024-10-25T17:05:19,602][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=24
[2024-10-25T17:05:21,037][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=25
[2024-10-25T17:05:23,282][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=26
[2024-10-25T17:05:24,850][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=27
[2024-10-25T17:05:27,728][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=28
[2024-10-25T17:05:28,313][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=29
[2024-10-25T17:05:28,607][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=30
[2024-10-25T17:05:29,197][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=31
[2024-10-25T17:05:33,938][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=32
[2024-10-25T17:05:34,086][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=33
[2024-10-25T17:05:36,126][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=34
[2024-10-25T17:05:37,291][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=35
[2024-10-25T17:05:37,343][INFO ][o.e.x.m.i.a.AdaptiveAllocationsScalerService] [runTask-0] adaptive allocations scaler: scaled [.elser-2-elasticsearch] to [2] allocations.
[2024-10-25T17:05:38,594][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=36
[2024-10-25T17:05:39,026][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=37
[2024-10-25T17:05:39,178][TRACE][o.e.x.i.s.e.ElasticsearchInternalService] [runTask-0] Executing batch index=38
(...)

```